services/app-api/utils/kafka/kafka-source-lib.ts

```diff
@@ -0,0 +1,200 @@
+    const { eventID, eventName } = record;
+    const dynamoRecord = {
+      NewImage: this.unmarshall(dynamodb.NewImage),
+      OldImage: this.unmarshall(dynamodb.OldImage),
+      Keys: this.unmarshall(dynamodb.Keys),
+    };
+    return {
+      key: Object.values(dynamoRecord.Keys).join("#"),
+      value: this.stringify(dynamoRecord),
+      partition: 0,
+      headers: { eventID: eventID, eventName: eventName },
+    };
+  }
+
+  async createS3Payload(record: S3EventRecord) {
+    const { eventName, eventTime } = record;
+    let entry = null;
+    if (!eventName.includes("ObjectRemoved")) {
+      entry = await s3Lib.get({
+        Bucket: record.s3.bucket.name,
+        Key: record.s3.object.key,
+      });
+    }
+
+    return {
+      key: record.s3.object.key,
+      value: entry,
+      partition: 0,
+      headers: { eventName, eventTime },
+    };
+  }
+
+  topic(t: string) {
+    if (this.version) {
+      return `${this.topicPrefix}.${t}.${this.version}`;
+    } else {
+      return `${this.topicPrefix}.${t}`;
+    }
+  }
+
+  async createOutboundEvents(records: any[]) {
+    let outboundEvents: { [key: string]: any } = {};
+    for (const record of records) {
+      let payload, topicName;
+      if (record["s3"]) {
+        // Handle any S3 events
+        const s3Record = record as S3EventRecord;
+        const key: string = s3Record.s3.object.key;
+        topicName = this.determineS3TopicName(s3Record.s3.bucket.arn);
+
+        // Filter for only the response info
+        if (
+          !topicName ||
+          !key.startsWith("fieldData/") ||
+          !key.includes(".json")
+        ) {
+          continue;
+        }
+        payload = await this.createS3Payload(record);
+      } else {
+        // DYNAMO
+        topicName = this.determineDynamoTopicName(
+          String(record.eventSourceARN.toString())
+        );
+        if (!topicName) continue;
+        payload = this.createDynamoPayload(record);
+      }
+
+      //initialize configuration object keyed to topic for quick lookup
+      if (!(outboundEvents[topicName] instanceof Object))
+        outboundEvents[topicName] = {
+          topic: topicName,
+          messages: [],
+        };
+
+      //add messages to messages array for corresponding topic
+      outboundEvents[topicName].messages.push(payload);
+    }
+    return outboundEvents;
+  }
+
+  async handler(event: any) {
+    if (!connected) {
+      await producer.connect();
+      connected = true;
+    }
+    console.log("Raw event", this.stringify(event, true));
+
+    if (event.Records) {
+      // if dynamo
+      const outboundEvents = this.createOutboundEvents(event.Records);
+
+      const topicMessages = Object.values(outboundEvents);
+      console.log(
+        `Batch configuration: ${this.stringify(topicMessages, true)}`
+      );
+
+      await producer.sendBatch({ topicMessages });
+    }
+
+    console.log(`Successfully processed ${event.Records.length} records.`);
```

Comment:
Improvement/Question: Should we have a structured logger across all of these services?

Reply:
I would appreciate that, but these services have historically been dumping console.log to the cloudwatch logs.